Add Brand resource with DTOs and data transformation

marierinden opened the pull request:
### Description

This pull request introduces functionality for managing brand-related operations with type-safe DTOs and data mapping tools. Key updates include:

- **BrandResource**: Implements CRUD operations for brands with validation and error handling.
- **BrandDataMapper**: Handles transformation of brand data between requests, responses, and DTOs.
- **BrandCollectionResponse**: A DTO for type-safe handling of brand collection responses.
- **BrandResponse**: A DTO encapsulating details of individual brand responses.
- **BrandRequest**: A DTO for safely handling incoming brand-related request data.

marierinden committed:
Add BrandRequest DTO for handling brand request data

This commit introduces a `BrandRequest` class implementing the `RequestInterface`. It includes properties for `name`, `description`, and an optional `iri`, alongside getter methods and a `toArray` method for serialization. This lays the groundwork for managing brand-related requests.

marierinden committed:
Add BrandResponse DTO class

Introduce the `BrandResponse` DTO to encapsulate brand-related response data. This class extends `AbstractResponse` and includes properties for `name` and `description` with corresponding getters.

marierinden committed:
Add BrandCollectionResponse DTO implementation

Introduces a DTO class for brand collection responses, extending the abstract collection response. This provides type safety and structure for handling brand-related data in the API responses.

marierinden committed:
Add BrandDataMapper to handle brand data transformation

Introduces BrandDataMapper implementing DataMapperInterface for converting brand-related request and response data. Includes methods to map single and collection responses into DTOs and convert requests into the appropriate format. Ensures error handling for data mapping failures.

marierinden committed:
Add BrandResource with CRUD operations

Introduce `BrandResource` to handle brand-related operations, including create, read, update, and delete functionality. This implementation ensures type safety, request validation, and error handling for seamless client interactions.

marierinden committed:
Update return type in BrandCollectionResponse::getMembers

Replaced the return type annotation from ResponseInterface to BrandResponse for better type clarity and accuracy. Removed an unused import statement to clean up the code.

marierinden committed:
Add missing newline for code readability consistency

Ensure consistent formatting by adding newlines between methods. This improves code readability and aligns with standard coding practices.

marierinden committed:
"Add strict types declaration to BrandRequest class

Declared strict types in BrandRequest to enforce type safety across the class. Removed redundant PHPDoc comments for constructor parameters as they are now self-explanatory."

marierinden committed:
Refactor BrandDataMapper to streamline annotations and imports

Removed unused imports and updated PHPDoc comments to improve clarity and consistency. Adjusted type hinting and exception annotations for better alignment with coding standards. No changes to functionality were made.

marierinden committed:
Refactor BrandResource endpoint and exception handling

Updated the API endpoint URL in BrandResource for consistency and replaced unqualified `ClientExceptionInterface` references with fully qualified names. This improves clarity and ensures proper exception documentation across methods.

fredrik-tveraaen-apiera commented on src/DTO/Request/Brand/BrandRequest.php:
```diff
@@ -0,0 +1,47 @@
+ */
+final readonly class BrandRequest implements RequestInterface
+{
+    public function __construct(
```
Missing store and image with getters

fredrik-tveraaen-apiera commented on src/DTO/Response/Brand/BrandCollectionResponse.php:
```diff
@@ -0,0 +1,26 @@
+
+use Apiera\Sdk\DTO\Response\AbstractCollectionResponse;
+
+/**
```
This need to be corrected.

fredrik-tveraaen-apiera commented on src/DTO/Response/Brand/BrandResponse.php:
```diff
@@ -0,0 +1,45 @@
+ * @author Marie Rinden <marie@shoppingnorge.no>
+ * @since 0.3.0
+ */
+final readonly class BrandResponse extends AbstractResponse
```
Missing store and image with getters

fredrik-tveraaen-apiera commented on src/DataMapper/BrandDataMapper.php:
```diff
@@ -0,0 +1,95 @@
+    {
+        try {
+            $members = array_map(
+                fn(array $brand): ResponseInterface =>
```
This could be added directly below. Look at AttributeDataMapper as reference

fredrik-tveraaen-apiera commented on src/Resource/BrandResource.php:
```diff
@@ -0,0 +1,165 @@
+ */
+final readonly class BrandResource implements RequestResourceInterface
+{
+    private const string ENDPOINT = '/brand';
```
Should be changed to /brands

fredrik-tveraaen-apiera commented on src/Resource/BrandResource.php:
```diff
@@ -0,0 +1,165 @@
+
+        /** @var BrandCollectionResponse $collectionResponse */
+        $collectionResponse = $this->mapper->fromCollectionResponse($this->client->decodeResponse(
+            $this->client->get(self::ENDPOINT, $params)
```
Should use store iri as prefix: $request->getStore() . self::ENDPOINT

fredrik-tveraaen-apiera commented on src/Resource/BrandResource.php:
```diff
@@ -0,0 +1,165 @@
+        $collection = $this->find($request, $params);
+
+        if ($collection->getTotalItems() < 1) {
+            throw new InvalidRequestException('No alternate identifier found matching the given criteria');
```
Change to brand

fredrik-tveraaen-apiera commented on src/Resource/BrandResource.php:
```diff
@@ -0,0 +1,165 @@
+        $requestData = $this->mapper->toRequestData($request);
+
+        /** @var BrandResponse $response */
+        $response = $this->mapper->fromResponse($this->client->decodeResponse(
```
Add store prefix: $request->getStore() . self::ENDPOINT

fredrik-tveraaen-apiera committed:
Add unit tests for BrandRequest DTO

This commit introduces comprehensive unit tests for the `BrandRequest` DTO. Tests cover object instantiation, constructor behavior with full and minimal parameters, property getters, and the `toArray` method to ensure expected output consistency.

fredrik-tveraaen-apiera committed:
Add unit tests for BrandResponse DTO

Implement comprehensive unit tests for the BrandResponse class, ensuring constructor, getters, and readonly properties are correctly defined and functioning. Covers tests for minimal and full parameters as well as interface and class type checks.

fredrik-tveraaen-apiera committed:
Add unit tests for BrandCollectionResponse

Introduce comprehensive unit test coverage for `BrandCollectionResponse`. Tests validate class instantiation, type inheritance, getters, and behavior with both minimal and complete input parameters. This ensures correctness and boosts confidence in its implementation.

fredrik-tveraaen-apiera committed:
Add unit tests for BrandDataMapper functionality

This commit introduces comprehensive unit tests for the BrandDataMapper class to ensure correct mapping of response and request data. It covers a variety of cases, including proper field mapping, handling nullable fields, empty collections, and invalid data scenarios, improving reliability and resilience.

fredrik-tveraaen-apiera committed:
Fix missing newline at end of file in test class

fredrik-tveraaen-apiera committed:
Add unit tests for BrandResource functionality

Introduce comprehensive unit tests for BrandResource, including methods for CRUD operations and data validation. These tests ensure proper behavior, handle edge cases, and validate exception throwing for invalid inputs.

fredrik-tveraaen-apiera committed:
Add integration tests for BrandResource functionality

This commit introduces integration tests covering BrandResource operations, including finding, creating, updating, and deleting brands. It ensures API requests are properly handled and validated, improving test coverage and reliability.

marierinden committed:
Add `store` and `image` fields to Brand entities

This update introduces `store` and `image` fields to `BrandRequest` and `BrandResponse` DTOs, ensuring these attributes are included in API interactions. Adjustments were also made to the `BrandResource` to use store-specific endpoints and ensure proper handling. These changes refine the structure of the brand-related entities and improve data completeness.

marierinden committed:
Make Brand fields nullable and add store validation.

Updated `BrandRequest` and `BrandResponse` to make `description`, `store`, and `image` nullable for better flexibility. Introduced a required validation for the `store` field in `BrandResource` to ensure consistency. Added `brand` handling in `ApieraSdk` with support for a new `BrandResource` and its associated data mapping.

fredrik-tveraaen-apiera committed:
Merge release/0.3.0: combine brand, distributor and file resources

fredrik-tveraaen-apiera committed:
Reorder properties in Brand DTOs and update tests.

Reordered `store` and `image` properties in `BrandRequest` and `BrandResponse` constructors for consistency. Updated corresponding unit tests to reflect these changes and ensure proper functionality. No functional changes were made to the application logic.

fredrik-tveraaen-apiera committed:
Add Brand management examples to documentation

This update enhances the documentation by including examples for managing brands such as finding, searching, creating, updating, and deleting brands. These additions offer clear guidance for developers integrating brand-related functionality.
